### ISSUE TITLE
Add support for per-authentication-flow state to SocialProviders (feature request/design proposal)

### ISSUE THREAD

**Issue by asazernik** [mohiva/play-silhouette#391]
The main application for this is redirection to the same page that a user originally requested but was not authenticated for; in OAuth2 there is a separate redirect_url (aka callbackURL) param, but most providers restrict that to a preset whitelist of URLs in order to combat CSRF attacks; the best practice recommended by e.g. [Google](https://developers.google.com/identity/protocols/OpenIDConnect#sendauthrequest) is to have `state` be a structured field that includes both the CSRF token _and_ "any other information needed to recover the context when the user returns to your application, e.g. the starting URL." Sessions and flash scope don't quite work, because they break in the case where a user has more than one tab/window open.

My proposal is to make two changes to the SocialProvider trait:
1. Add a `userState: Map[String, String]` param to `SocialProvider.authenticate[B]()`, which will be inserted into the auth redirect on the initial request (i.e. when `authenticate()` returns a `Left[Result]`) and be re-inserted into the authInfo in a provider-specific way on auth completion.
2. Add a `retrieveUserState(authInfo: A): Map[String, String]` which will allow a caller to retrieve this per-flow state on a successful authentication.

I have a fairly good grasp on how to implement this for OAuth2Provider - change `authenticate()` to combine both `serializedState` (which I'd probably rename to `csrfState` for clarity) and the new `userState` param to get the `state` auth URL parameter, and change the `StateProvider.validate[B]` method to take an explicit `providerState` parameter instead of reading out the whole `state` callback URL param.

I would appreciate both comments on this design as it applies to OAuth2 and tips for how to implement this for other providers; once I've gotten some design feedback I'll start putting a PR together.

**Comment by akkie** [comment]
I had a length [discussion](https://groups.google.com/forum/#!topic/play-silhouette/vBQXOTcD1AU) about this issue with @kajmagnus a while back. Maybe you can find some design decisions in this discussion.

Some notes to your proposal:

* The state provider isn't only meant for CSRF protection. It can also be used for any other state parameter. The only missing part is the context bound state, because the state provider is injected on application instantiation and therefore you cannot provide state params per request. Maybe this could be done with a callback function which can be registered for a state provider.

* The OAuth2 provider is the only provider who supports this state parameter. So it seems a bit odd to add the state parameter to the other providers.

* How does the `retrieveUserState` method retrieve the state?

**Comment by asazernik** [comment]
Addressing your notes in order:

* The API for OAuth2StateProvider is actually inconsistent on this point. When the state is being sent to the provider, it just hands a string to the OAuth2Provider code (through State.serialize) and only the provider code knows that this is transmitted to the provider using the `state` param; but on the receive end, the OAuth2StateProvider is expected to have code to extract the state param (e.g. CookieStateProvider.providerState). As part of this change, I would like to change the StateProvider trait to *only* make its implementations deal with: a) finding a secure place to store a CSRF token, and retrieving it for cross-checking, and b) serializing that CSRF token so the OAuth2Provider can stuff it into the redirect URL, and unserializing it when the OAuth2Provider hands it back. This makes the name OAuth2CSRFChecker more fitting, but that's kind of true already.

* The state parameter is the OAuth2 way of dealing with this problem, but the problem is general to all SocialProviders. I'd like to change the SocialProvider API to support this generally, and make the state parameter be OAuth2's implementation, but other SocialProviders might implement this very differently (hence my request for ideas about how to pull this off for e.g. OpenIDProvider). In order to do this incrementally, starting *only* with OAuth2Provider, it might be better to extend SocialProvider with a StatefulSocialProvider trait like so:

```scala
trait StatefulSocialProvider extends SocialProvider {

    // makes sure to be cheap in the empty-map case; userState ignored if the request indicates completed auth
    abstract def authenticate[B](userState: Map[String, String])(implicit request: ExtractableRequest[B]): Future[Either[Result, A]]

    override def authenticate[B]()(implicit request: ExtractableRequest[B]) = authenticate(Map.empty())

    // abstract
    def retrieveUserState(authInfo: A): Map[String, String]
}
```

* In the OAuth2Provider case, OAuth2Info would need an additional userState field containing non-CSRF-token parts of the `state` parameter from the second (redirected-from-provider) request. This would be populated in authenticate(). Then when `retrieveUserState` just extracts that from the AuthInfo. It's only a method of the provider so that - like `retrieveProfile` - it can be called by users without writing provider-type-specific code.

I read over your exchange with @kajmagnus, and my proposal looks a lot like his first API proposal; I prefer this to his second proposal (adding a generator function as a param to the Provider classes) because it:

* keeps auth-flow logic centralized in the controller, not split between the controller and the DI module, which in my very subjective opinion is more readable.

* allows data to be stuffed into the state which are not trivially derivable from the request, or which may differ depending on which auth endpoint a client hits.

**Comment by akkie** [comment]
If you find a suitable solution for all provides then I'm fine with your suggestion. I'm also fine with any other elegant, abstract solution. But changing the main API for all Social providers, because of the functionality of a single provider would be a very odd solution.

**Comment by akkie** [comment]
@asazernik Any progress on this issue?

**Comment by asazernik** [comment]
Not at this point; my plan is to get started on this towards October or so, when a project of mine gets to a stage where I can use it (and test it).

**Comment by asazernik** [comment]
Hi there @akkie! Finally back onto this. So much for "October". But now my work schedule is a bit more on track.

So! Looking harder at the overall problem, it seems that the difficulty in figuring out an API for this functionality springs from an unusual choice the SocialProvider API makes - combining the initiation of the flow with the future requests involved. There's really a difference between the two, in that in initiation we're taking a request that probably has nothing to do with the auth protocol and just redirecting properly into the auth flow. In the latter case, we're continuing the back-and-forth of the auth protocol until we get an AuthInfo Arbitrary user state (e.g. a post-auth redirect URL) only really makes sense when the caller distinguishes between the two cases - callers want to pass state into the original flow at initiation only, and get back that state when they successfully retrieve an AuthInfo on the same auth attempt (browser tab, basically).  This matches pretty roughly with the the flow that other OAuth{1,2} implementations impose; I haven't gotten around to looking at OpenID libraries, but I'd assume they're similar.

So I do think an API change is necessary, but only for users that actually want the user state functionality; if they don't, the authenticate() method can from my quick inspection of the existing SocialProvider implementations be implemented in terms of the other two.

For quick review of the API before I get too deep into the weeds on this, I've uploaded [a branch](https://github.com/asazernik/play-silhouette/tree/per-auth-attempt-state) containing the new SocialProvider trait and an example of the splitting of `authenticate()` for OAuth2. From my quick look at the other two direct implementations of SocialProvider (OpenIDProvider and OAuth1Provider) this approach will be pretty simple in those other cases. Let me know what you think.

Once I have that split done, the implementations of saving user state in each protocol should (*crosses fingers*) be fairly simple.

**Comment by akkie** [comment]
@asazernik Sorry, but I do not find the time in the next weeks to deal with this issue. The change would be to big for the final 4 release.

**Comment by asazernik** [comment]
Yeah, definitely not the kind of change that should go into an already-in-beta major version :-) 

Just wanted to get the design that's in my head out into the world while *I* have time.

**Comment by JamesSullivan** [comment]
@asazernik If there is anything I can help with let me know. I definitely think it is a useful feature for a seamless end user experience.

**Comment by asazernik** [comment]
@JamesSullivan I have what I think is a complete version of the code changes done - the diff is only around 200 lines - though I still need to write tests.

Mostly what I need is review and testing. I have a project right now using only OAuth2, but the changes are also to OpenID and OAuth1. If you have any code (preferably in a non-mission-critical test environment) using either of those, I'd appreciate if you can integrate my changes into your silhouette installation and let me know what breaks and what's annoying about the API.

My branch is updated at the [same location as before](https://github.com/asazernik/play-silhouette/tree/per-auth-attempt-state).

**Comment by JamesSullivan** [comment]
@asazernik Unfortunately my projects only use OAuth2 but I will look around to if I can find something with OpenID and OAuth1. Interestingly (for me at least as I know very little about authentication) I noticed Facebook maintains referer so this is not really necessary if you are only using Facebook. I tried this with Google Identity but I have done something incorrectly and it did not work. It seems very simple (include the url in  
`socialProvider.initiateAuthentication(Some(uri))`

 and retrieve it during the callback
```
socialProvider.continueAuthentication().flatMap {
          case Right(authInfo) => println("userState "  + authInfo.userState)
```
but I always receive none for the userState. I will try again when there are tests so I can make sure I am doing it correctly.

**Comment by asazernik** [comment]
That does not surprise me; I have an OAuth2 application that I *could* test this on, but I have in fact done zero testing so far :-P

I'll work on some sanity-check unit tests this weekend.

**Comment by asazernik** [comment]
(And yeah, that was the API usage I'd intended; good to know that it's reasonably intuitive to someone other than me.)

**Comment by ilucker** [comment]
Here is how i solve this in my projects: https://gist.github.com/ilucker/142bb4eac40221cdb857

**Comment by asazernik** [comment]
@JamesSullivan Okay, two months on I'm spending another weekend on this. And saw that I asked you to test it, when I'd only changed the prototypes but hadn't actually **implemented** any of the functionality yet.

:grimacing:

I'll poke you soon about actual testing.

**Comment by akkie** [comment]
I've changed my opinion about the need to implement the state functionality into every provider. Every provider has different needs so it doesn't make sense to find a workaround for other providers only to keep the API in sync. It makes sense to provide a unique interface for all providers for the common functionality. We have this currently with the `authenticate` method. But for the additional functionality we should provide a different interface for every provider. Maybe it makes sense to create a `SocialStateProvider` trait which contains the API to handle user state. So it can be reused for providers which supports also user state. I think OpenID Connect would be such a candidate. The OAuth2Provider depends then directly on `SocialStateProvider` instead of `SocialProvider`.

```scala
case class StatefulResult(result: Result, state: Option[Map[String, String]])
trait SocialStateProvider extends SocialProvider {
  def authenticate[B](state: Option[Map[String, String]])(implicit request: ExtractableRequest[B]): Future[Either[StatefulResult, A]]
}
```

@asazernik @todor-kolev Any suggestions?

**Comment by todor-kolev** [comment]
I reckon we will also have to change the `OAuth2StateProvider#build` method signature to include the `state: Option[Map[String, String]]` param??

**Comment by akkie** [comment]
@todor-kolev Yes, I think this would be the best solution :+1:

**Comment by Saheb** [comment]
Anyone working on this one?

**Comment by Saheb** [comment]
@akkie - Is anyone working on this? If not, shall I take this up?

**Comment by akkie** [comment]
I do not work on this. It would be great if you could provide a PR!

**Comment by Saheb** [comment]
Let me try to come with some working solution.

**Comment by akkie** [comment]
@Saheb Would be cool if you could consider the solution from my [comment](https://github.com/mohiva/play-silhouette/issues/391#issuecomment-252409259) :wink:

**Comment by Saheb** [comment]
@akkie did read the conversation above - need to deep dive in to figure out what actually makes sense.

**Comment by akkie** [comment]
Fixed with #503. Thanks to @Saheb :+1:
